### PR TITLE
Feature: use git branches/tags as versions

### PR DESCRIPTION
**Pull request by becker33**
Building on #24639, this allows versions to be prefixed by `git.`. If a version begins `git.`, it is treated as a git ref, and handled as git commits are starting in the referenced PR.

An exception is made for versions that are `git.develop`, `git.main`, `git.master`, `git.head`, or `git.trunk`. Those are assumed to be greater than all other versions, as those prefixed strings are in other contexts.

TODO:
- [ ] docs (could be combined with the git commit docs PR)

### CODE REVIEW

**Comment by becker33** [comment]
@haampie you might be happy about some of the refactoring in this PR.

**Comment by psakievich** [comment]
@becker33 I'm taking a look at this now. Does this have sufficient overlap to merge in the concepts from #30998?  Or rather is there room to add it. It seems like my draft PR will need to be totally re-done after this merges anyways.

The key test for #30998 which I'm still failing on is this:

`spack solve nalu-wind ^trilinos@[hash]` and the reason why it fails is that `nalu-wind` develops on the highest precedent infinity version `trilinos@develop`. My PR should fix that issue, but there is something going on in the concretizer that I don't fully understand yet.

The other main feature I want is to be able to skip the commit lookup all together by assigning the version.  The clone and lookup procedure really slows down my concretization. This feature will work without it, but is less than optimal. We can chat via slack if that is helpful.

If this is too much then I can just refactor when the time comes.

**Comment by psakievich** [comment]
Ooo it does take care of my test case! Just tested it!
`nalu-wind ^trilinos@git.[hash]`

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py]: This introduces a redundant copy everywhere `Version` is called (which is a lot of places) -- should check with `isinstance()` and not bother constructing a `str` if we do not need to.

CPython *can* intern strings so it might optimize this out in a lot of cases, but it doesn't always.

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py]: see above -- I think this should check with `isinstance` and not copy unconditionally.

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py:73]: docstring needs updating.

File: lib/spack/spack/solver/asp.py
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py:1435]: can you put a comment here saying why this doesn't matter for git versions?  I had to think about this for a minute to work out why we don't care.

File: lib/spack/spack/test/versions.py
Comment (tgamblin) [inline lib/spack/spack/test/versions.py:566]: There should be another simple test to verify that `Version()` returns the right thing with commit, branch, and plain version inputs.

File: lib/spack/spack/test/versions.py
Comment (tgamblin) [inline lib/spack/spack/test/versions.py:648]: what's holding this back on windows?  Seems like if we have `git` in the path it should be fine.

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py:177]: Doesn't it make sense to also allow `git.<commit hash>` for consistency?  I can see people trying to make commits that way based on branches and failing mysteriously.

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py:327]: It doesn't look like there is an implementation of `up_to` in `GitVersion` that preserves this behavior -- should there be?

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py:455]: So happy this is separate now.  But needs a docstring.

File: lib/spack/spack/version.py
Comment (tgamblin) [inline lib/spack/spack/version.py]: it looks like this allows commit versions to start with `git.`, but `Version()` doesn't.  Need a test to ensure that both work.

Comment (tgamblin) [reply lib/spack/spack/version.py]: In other version types, `string` preserves the *exact* original string, but here we lose `git.`.  I actually think it makes sense to keep that... shouldn't we?

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/fetch_strategy.py
Comment (psakievich) [inline lib/spack/spack/fetch_strategy.py]: How does this behave in the case that `isinstance(version, spack.version.GitVersion) and !hasattr(pkg, "git")`? Should we add a check and raise an exception in that case?.

Comment (psakievich) [reply lib/spack/spack/fetch_strategy.py]: Also I think this will fail for `git.master` versions. The `kwargs` for the fetch strategy are hard coded to use `commit` below this.

Should add tests for the different `git.[ref types]` that are supported.

File: lib/spack/spack/solver/asp.py
Comment (psakievich) [inline lib/spack/spack/solver/asp.py]: The name `generate_commit_lookup` seems to be a little too specific for the git variations now supported by this PR.  It might be better to rename it or add another layer of abstraction to account for ref types such as `git.branch` and `git.tags`

File: lib/spack/spack/version.py
Comment (psakievich) [reply lib/spack/spack/version.py]: IMO it makes sense to maintain the exact string the user specified for the spec in the version. This will make it clearer when reviewing a list of specs from `spack find`

**Review by psakievich — CHANGES_REQUESTED** [review]
Looks really good. There are a few holes regarding git refs other than commits though.  I pointed out at least one in my review, but I think more tests need to be added for the various types of refs supported. I pulled this down and tested staging a `git.[branch]` locally.
```
spack stage yaml-cpp@git.master
From https://github.com/jbeder/yaml-cpp
 * branch            HEAD       -> FETCH_HEAD
==> Warning: There is no checksum on file to fetch yaml-cpp@git.master safely.
==>   Fetch anyway? [y/N] y
error: pathspec 'git.master' did not match any file(s) known to git
==> Error: All fetchers failed for spack-stage-yaml-cpp-git.master-hmbhd2gnl2342shpta7dhprcajxagslg
```

`spack solve yaml-cpp@git.master` works fine.

File: lib/spack/spack/version.py
Comment (becker33) [reply lib/spack/spack/version.py:177]: This does allow `git.<commit hash>`. The first check is for `git.`, we only check the contents if it doesn't start with that string.

File: lib/spack/spack/test/versions.py
Comment (becker33) [reply lib/spack/spack/test/versions.py:648]: I don't know what is holding it back, but the previous test `test_git_commit_comparisons` was marked with the same skip, and they follow almost-identical code paths so I'm pretty sure the same reasoning will apply.

File: lib/spack/spack/version.py
Comment (becker33) [reply lib/spack/spack/version.py]: `Version()` already does allow `git.<commit hash>`. The test I've added confirms that.

The reason I have `self.string` strip `git.` out and `str(self)` include it is that we often need the git ref alone. I can change that to be some other attribute and keep the full string in `self.string` if you prefer.

File: lib/spack/spack/version.py
Comment (becker33) [reply lib/spack/spack/version.py:73]: Tom also has an idea to represent all versions in the same triply nested tuple syntax to eliminate the need for `@coerced` that I'm quite excited about. Separate PR though.

File: lib/spack/spack/version.py
Comment (becker33) [reply lib/spack/spack/version.py:327]: This is the code for `satisfies`, which does have a counterpart. The code for `up_to` didn't need to change.

File: lib/spack/spack/version.py
Comment (tgamblin) [reply lib/spack/spack/version.py]: Yeah I think it's better to be consistent with other versions and keep the original input string for output.

File: lib/spack/spack/version.py
Comment (tgamblin) [reply lib/spack/spack/version.py:73]: yeah I talked about that with Tom and it's way way faster.  Definitely separate PR though.

File: lib/spack/spack/fetch_strategy.py
Comment (becker33) [reply lib/spack/spack/fetch_strategy.py]: Modified it to raise if there is no 'git' attribute.

Treating branches/tags as commits wasn't going to cause a problem, it just has more conservative fetch semantics. I've changed it to separate by ref type.

**Comment by becker33** [comment]
@tgamblin @psakievich I believe I've addressed all review comments.

File: lib/spack/spack/version.py
Comment (psakievich) [inline lib/spack/spack/version.py]: I was playing around with this a little more, and I think it might be a good time to add a custom error. If I knowingly give it a wrong git ref (but a not unreasonable one) we get a rather unclear error message:

```console
$ spack stage yaml-cpp@git.develop
From https://github.com/jbeder/yaml-cpp
 * branch            HEAD       -> FETCH_HEAD
fatal: Not a valid object name develop^{commit}
==> Error: Command exited with status 128:
'/usr/bin/git' '-c' 'advice.detachedHead=false' 'cat-file' '-e' 'develop^{commit}'
```

I don't feel like this is a show stopper though so I won't request holding up the PR over it.  However, I think if not in this PR then soon. 🙂

**Review by psakievich — DISMISSED** [review]
LGTM

File: lib/spack/spack/version.py
Comment (becker33) [reply lib/spack/spack/version.py]:
```
$ spack stage yaml-cpp@git.develop
==> Error: develop is not a valid git ref for yaml-cpp
```
better?

Comment (psakievich) [reply lib/spack/spack/version.py]: 💯

File: lib/spack/spack/fetch_strategy.py
Comment (trws) [inline lib/spack/spack/fetch_strategy.py:1578]: Missed it was a subclass.

Comment (becker33) [reply lib/spack/spack/fetch_strategy.py:1578]: GitVersion inherits from VersionBase

**Review by psakievich — DISMISSED** [review]

File: lib/spack/spack/fetch_strategy.py
Comment (trws) [reply lib/spack/spack/fetch_strategy.py:1578]: Yup, that's why I resolved it and added the comment to that effect.

File: lib/spack/spack/directives.py
Comment (tgamblin) [inline lib/spack/spack/directives.py]:
```python
pkg.versions[version] = kwargs
```

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
